### DISCUSSION TITLE
[Custom Source] Parsing JSON in to cmp readable table.

### DISCUSSION THREAD

**Discussion by rudfol** [hrsh7th/nvim-cmp#1531] [Q&A] [answered]
Hi,

The difficulty I have is understanding what kind items cmp accepts. One item would be 'label' which I understand to be
 used as an entry in the completion window.
To create a custom source I've used the template you provided.

![boom 3](https://user-images.githubusercontent.com/6250738/233870384-e8fa7106-b62e-440d-9460-1d49322d5762.PNG)

The format of this JSON file I copied from the [rafamadriz ](https://github.com/rafamadriz)/[friendly-snippets](https://github.com/rafamadriz/friendly-snippets) json files inside snippet folder.
![1 boom](https://user-images.githubusercontent.com/6250738/233869453-034858c9-7fc6-4738-ae5d-0fc0094e1483.PNG)

So far the result I get is nothing. 👎  

Expected results would be  'MediaItem reaper.AddMediaItemToTrack'
(residing in the first curly bracket) 
to be used as a title on the documentation window, 

contents of 'prefix:' for label, 

what's in the 'body' for completion (snippetable) and 

the content of 'description:' for documentation.

How should I go about doing this please?

P.S I have already added the source to the cmp config and confirmed that the default template you provided for 'month' works.

**Comment by rudfol** [comment]
So far I've read multiple cmp sources and from what I understand below should work, just absolutely no idea why it does not however.
![1 boom](https://user-images.githubusercontent.com/6250738/234093788-119ff16d-a26c-4b87-a700-40bddcab56b3.PNG)

I will be continuing to work on this, but I would loveee a pointer in the right direction please.

**Comment by rudfol** [comment, marked as answer]
For anyone looking for answers:
![1 boom2](https://github.com/hrsh7th/nvim-cmp/assets/6250738/6749d979-6b44-4447-a205-89957b903e0f)

'- #input' <-- is for replacing what you already typed with a completion rather than completion concatenating on to what you typed.

the definition of things like 'label' and 'InsertTextMode' go see LSP Spec at: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/

**Comment by joe-dil** [comment]
I created a plugin to parse JSON and extract fields to be used for completions, check it out joe-dil/json-cmp

**Reply by rudfol** [reply]
that's pretty awesome, because it looks so easy to use. Cheers for the work on that

**Reply by rudfol** [reply]
I am back to setting up my own source with blink.cmp after many years and dayuuum, i don't even know what the hell is going on. I forgot all of it loool

**Comment by kinthaiofficial** [comment]
Writing a custom nvim-cmp source is well-documented once you find the right examples. The key is the source object interface:

```lua
local source = {}

-- Required: return source name
source.get_trigger_characters = function()
  return { '.', ':' }
end

-- Required: provide completion items
source.complete = function(self, params, callback)
  local items = {}
  
  -- params.context has cursor position, buffer info, etc.
  local bufnr = params.context.bufnr
  
  -- Build your items
  table.insert(items, {
    label = 'my_completion',
    kind = require('cmp').lsp.CompletionItemKind.Text,
    detail = 'My custom completion',
    documentation = {
      kind = 'markdown',
      value = '**Description**\n\nThis is my item'
    },
  })
  
  callback({ items = items, isIncomplete = false })
end

-- Register the source
require('cmp').register_source('my_source', source)
```

**For JSON parsing to cmp table**:
```lua
source.complete = function(self, params, callback)
  local json_str = vim.fn.system('your-command-that-returns-json')
  local ok, data = pcall(vim.json.decode, json_str)
  
  if not ok then
    callback({ items = {}, isIncomplete = false })
    return
  end
  
  local items = vim.tbl_map(function(entry)
    return {
      label = entry.name,
      detail = entry.type,
      documentation = entry.description,
    }
  end, data)
  
  callback({ items = items })
end
```

**Async sources**: for slow data (API calls, filesystem), use `callback` asynchronously — don't block the complete function. Use `vim.schedule` or `plenary.async` to keep the event loop responsive.

**Sorting/filtering**: cmp handles fuzzy matching on labels automatically. You just need to return the candidate set.

What's the data source for your JSON — an external API, a local file, or a tool output?
